### PR TITLE
feat(router): add expand attempts support in payments retrieve response

### PR DESCRIPTION
**Pull request by sai-harsha-vardhan**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Currently we only have active payment attempt details in payments retrieve response. For retries we would have multiple attempts, need an option to see all the attempts linked to payment intent. 

Add `expand_attempts` in payments retrieve request, when enabled gives list of attempts linked to payment intent in response.

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
Tested Manually
image

image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/core/payments/transformers.rs [hunk 1]
```diff
@@ -292,6 +294,21 @@ where
...
+                Some(
+                    attempts
+                        .into_iter()
+                        .map(ForeignInto::foreign_into)
+                        .collect(),
+                )
+            }
+        }
+        None => None,
+    };
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/transformers.rs:311]:
```suggestion
    let attempts_response = option_attempts.map(|attempts| attempts.into_iter().map(ForeignInto::foreign_into).collect())
    };
```

File: crates/api_models/src/payments.rs [hunk 2]
```diff
@@ -293,6 +293,53 @@ pub struct PaymentsRequest {
...
+    #[schema(value_type = Option<CaptureMethod>, example = "scheduled")]
+    pub capture_method: Option<enums::CaptureMethod>,
+    /// The transaction authentication can be set to undergo payer authentication. By default, the authentication will be marked as NO_THREE_DS
+    #[schema(value_type = Option<AuthenticationType>, example = "no_three_ds", default = "three_ds")]
+    pub authentication_type: Option<enums::AuthenticationType>,
+    /// If the payment was cancelled the reason provided here
+    pub cancellation_reason: Option<String>,
+    /// A unique identifier to link the payment to a mandate, can be use instead of payment_method_data
+    pub mandate_id: Option<String>,
+    /// /// If there was an error while calling the connectors the code is received here
```
Comment (Narayanbhat166) [inline crates/api_models/src/payments.rs:329]:
```suggestion
    /// If there was an error while calling the connectors the code is received here
```

File: crates/api_models/src/payments.rs [hunk 3]
```diff
@@ -293,6 +293,53 @@ pub struct PaymentsRequest {
     pub feature_metadata: Option<FeatureMetadata>,
 }
 
+#[derive(
+    Default, Debug, serde::Serialize, Clone, PartialEq, ToSchema, router_derive::PolymorphicSchema,
+)]
+pub struct PaymentAttemptResponse {
+    /// Unique identifier for the attempt
+    pub attempt_id: String,
```
Comment (jarnura) [inline crates/api_models/src/payments.rs:301]: the key name will be `reference_id` and not `attempt_id`. It can be sometimes not attempt_id too

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/api_models/src/payments.rs [hunk 4]
```diff
@@ -293,6 +293,56 @@ pub struct PaymentsRequest {
...
+    /// The status of the attempt
+    #[schema(value_type = AttemptStatus, example = "charged")]
+    pub status: enums::AttemptStatus,
+    /// The payment attempt amount. Amount for the payment in lowest denomination of the currency. (i.e) in cents for USD denomination, in paisa for INR denomination etc.,
+    pub amount: i64,
+    /// The currency of the amount of the payment attempt
+    #[schema(value_type = Currency, example = "usd")]
```
Comment (Narayanbhat166) [inline crates/api_models/src/payments.rs:308]:
```suggestion
    #[schema(value_type = Option<Currency>, example = "usd")]
```

File: crates/api_models/src/payments.rs [hunk 5]
```diff
@@ -293,6 +293,56 @@ pub struct PaymentsRequest {
...
+    pub currency: Option<enums::Currency>,
+    /// The connector used for the payment
+    pub connector: Option<String>,
+    /// If there was an error while calling the connector the error message is received here
+    pub error_message: Option<String>,
+    /// The payment method that is to be used
+    #[schema(value_type = PaymentMethod, example = "bank_transfer")]
```
Comment (Narayanbhat166) [inline crates/api_models/src/payments.rs:315]:
```suggestion
    #[schema(value_type = Option<PaymentMethod>, example = "bank_transfer")]
```

**Review by Narayanbhat166 — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Pull request merged**